src/emulator/host/src/app.cpp

```diff
@@ -235,3 +247,74 @@ bool load_app(Ptr<const void> &entry_point, HostState &host, const std::wstring
 
     return true;
 }
+
+ExitCode load_app(Ptr<const void> &entry_point, HostState &host, const std::wstring &path, bool is_vpk) {
+    if (!load_app_impl(entry_point, host, path, is_vpk)) {
+        std::string message = "Failed to load \"";
+        message += wide_to_utf(path);
+        message += "\"";
+        message += "\nSee console output for details.";
+        error_dialog(message.c_str(), host.window.get());
+        return ModuleLoadFailed;
+    }
+
+    return Success;
+}
+
+ExitCode run_app(HostState &host, Ptr<const void> &entry_point) {
+    const CallImport call_import = [&host](uint32_t nid, SceUID main_thread_id) {
+        ::call_import(host, nid, main_thread_id);
+    };
+
+    const SceUID main_thread_id = create_thread(entry_point, host.kernel, host.mem, host.io.title_id.c_str(), SCE_KERNEL_DEFAULT_PRIORITY_USER, SCE_KERNEL_STACK_SIZE_USER_MAIN, call_import, false);
+    if (main_thread_id < 0) {
+        error_dialog("Failed to init main thread.", host.window.get());
+        return InitThreadFailed;
+    }
+
+    const ThreadStatePtr main_thread = find(main_thread_id, host.kernel.threads);
+    Ptr<void> argp = Ptr<void>();
+    if (!strncmp(host.kernel.loaded_modules.begin()->second->module_name, "SceLibc", 7)) {
+        const SceUID libc_thread_id = create_thread(host.kernel.loaded_modules.begin()->second->module_start, host.kernel, host.mem, "libc", SCE_KERNEL_DEFAULT_PRIORITY_USER, SCE_KERNEL_STACK_SIZE_USER_DEFAULT, call_import, false);
+        const ThreadStatePtr libc_thread = find(libc_thread_id, host.kernel.threads);
+        run_on_current(*libc_thread, host.kernel.loaded_modules.begin()->second->module_start, 0, argp);
+        libc_thread->to_do = ThreadToDo::exit;
+        libc_thread->something_to_do.notify_all(); // TODO Should this be notify_one()?
+        host.kernel.running_threads.erase(libc_thread_id);
+        host.kernel.threads.erase(libc_thread_id);
+    }
+
+    if (start_thread(host.kernel, main_thread_id, 0, Ptr<void>()) < 0) {
+        error_dialog("Failed to run main thread.", host.window.get());
+        return RunThreadFailed;
+    }
+
+    return Success;
+}
+
+void no_fb_fallback(HostState &host, GLuint *fb_texture_id) {
+    glGenTextures(1, fb_texture_id);
+    glClearColor(1.0, 0.0, 0.5, 1.0);
+    glClearDepth(1.0f);
+    glViewport(0, 0, host.display.image_size.width, host.display.image_size.height);
+    glMatrixMode(GL_PROJECTION);
```

Comment:
These matrix functions aren't in OpenGL 4.1 Core -- we only use shaders so they're not necessary.

Reply:
I removed those, `glLoadIdentity` and `glOrtho`.